### PR TITLE
test(e2e): add upgrade k0s tests for zero-downtime upgrade

### PR DESCRIPTION
**Pull request by bhatturu**
# PR: test(e2e): add upgrade k0s tests for zero-downtime upgrade

## What this does

Adds e2e coverage proving that a rolling Spur upgrade (daemon restarts simulating `rolling_upgrade.yml`) does **not** disrupt a managed k0s cluster running without WireGuard.

Today the rolling upgrade playbook restarts `spurctld` and `spurd` one node at a time, but nothing proves k0s survives the restart. k0s runs as a systemd unit independent of the spur daemons — the controller's Raft-replicated `K0sClusterState` must recover from the log on restart, and spurd's `ClusterSupervisor` must re-adopt the running k0s unit instead of restarting it. This test pins those invariants down.

The four upgrade scenarios and their coverage:

| Scenario | Coverage |
|----------|----------|
| Spur native (no WG, no k0s) | `TestAgentRestartPreservesRunningJob` in `test_deregistration.py` |
| Spur native (no WG) + managed k0s (no WG) | **this PR** |
| Spur WG + managed k0s on WG | not yet covered |
| Spur WG only (no k0s) | not yet covered |

## Approach

**New file**: `tests/native_host/e2e/test_upgrade_k0s.py`

**Fixture**: `k0s_native_cluster` — 3-node native k0s cluster (kuberouter CNI, rootful, no WireGuard, `allow_root_jobs=true`). Resets k0s state on setup and teardown. Starts daemons with `cluster.enabled=True` but does not call `k8s up` — the test controls timing.

**One test class** (`TestRollingUpgradePreservesK0s`) running the full upgrade sequence in a single test to avoid tearing down and re-provisioning k0s between steps:

1. `k8s up` → wait for spur-level `ready` phase (same gate as `test_k8s_multicp`)
2. Restart `spurctld` → assert k0s phase stays `ready`, membership unchanged, spur tracks all nodes
3. Restart `spurd` on each node one at a time → same assertions after all restarts
4. `k8s down` → `wait_ready` → submit a batch job → verify it completes

Step 4 calls `k8s down` before the batch job because `k8s up` reserves all scoped nodes for kubernetes, making them unavailable for batch scheduling (this is the intended behavior, tested by `test_k8s_scheduling`).

No changes to `cluster.py` or `conftest.py`.

### Assertions

1. k0s reaches `ready` over native (kuberouter) networking
2. Controller restart does not change k0s phase or membership
3. Controller restart does not deregister nodes from spur
4. Per-node rolling agent restart does not change k0s phase or membership
5. Each restarted agent re-registers with the controller within 60s
6. After `k8s down`, all nodes return to `idle` state
7. A new batch job completes after the full upgrade cycle

### CI

The `k0s` marker ensures CI's e2e workflow picks up this test and runs it on its 3-node VM cluster with proper bridged networking.

### CODE REVIEW

File: tests/native_host/e2e/test_upgrade_k0s.py [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+
+    def test_controller_restart_preserves_k0s(self, k0s_native_cluster):
+        c = k0s_native_cluster
+        cp_node = c.node_names[0]
+        workers = c.node_names[1:]
+
+        c.k8s_up(["--control-plane-node", cp_node])
+        c.wait_k8s_phase("ready", timeout=600)
+        _wait_condition(lambda: _k8s_nodes_ready(c, workers),
+                        timeout=480, desc="all workers Ready before upgrade")
```
[the remark this answers is not in the document]
Comment (bhatturu) [reply tests/native_host/e2e/test_upgrade_k0s.py]: **Acknowledged.** This is an infra limitation of cold QEMU VMs, not a test bug — the existing `test_wg_k0s.py` and `test_k8s_multicp.py` have the same dependency on pre-cached images and work in CI because the e2e VM images ship with k0s images pre-baked. Pre-pulling k0s system images requires a running containerd (only available after k0s starts), creating a chicken-and-egg problem. The `_k8s_nodes_ready` gate now checks **all nodes** (CP + workers) instead of only workers, so a NotReady control-plane is also caught.

File: tests/native_host/e2e/test_upgrade_k0s.py [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+        script = c.write_file("post-upgrade-job.sh", "#!/bin/bash\necho UPGRADE_OK\n",
+                              all_nodes=True)
+        out = c.sbatch(["--job-name=post-upgrade", script])
+        job_id = parse_job_id(out)
+        assert job_id is not None, f"sbatch did not return a job id: {out}"
+
+        state = wait_job(c, job_id, timeout=120)
+        assert state == "COMPLETED", (
+            f"post-upgrade job {job_id} ended in state {state}, expected COMPLETED"
+        )
```
[the remark this answers is not in the document]
Comment (bhatturu) [reply tests/native_host/e2e/test_upgrade_k0s.py]: **Fixed.** Changed assertion to `assert state in ("CD", "GONE")`, matching the pattern in `test_single_node.py`.

File: tests/native_host/e2e/test_upgrade_k0s.py [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+
+        output = c.read_output_on_any_node(f"slurm-{job_id}.out")
+        assert "UPGRADE_OK" in output, (
```
[the remark this answers is not in the document]
Comment (bhatturu) [reply tests/native_host/e2e/test_upgrade_k0s.py]: **Fixed.** Now passes an explicit `-o` path (`{c.remote_dir}/post-upgrade.out`) and reads from that, matching the pattern in `test_single_node.py`.

Also removed unused `re` import and `_IPV4_RE` constant caught by cleanup pass.

File: tests/native_host/e2e/test_upgrade_k0s.py [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+
+def _assert_k0s_healthy(c: SpurCluster, worker_names: list[str],
+                        canary_pod: str | None = None) -> None:
+    """Assert k0s cluster health: phase=ready, all nodes Ready, canary Running."""
+    status = c.k8s_status()
+    assert "ready" in status.lower(), f"k0s phase is not ready:\n{status}"
+
+    assert _k8s_nodes_ready(c, worker_names), (
+        f"not all k8s nodes are Ready:\n{_kubectl(c, 'get nodes --no-headers')}"
+    )
```
[the remark this answers is not in the document]
Comment (bhatturu) [reply tests/native_host/e2e/test_upgrade_k0s.py]: **Fixed.** `_assert_k0s_healthy` now passes `c.node_names` (all nodes including CP) instead of `worker_names`. All `_wait_condition` callers also switched from `workers` to `c.node_names`. A NotReady control-plane now fails the assertion.

**Review by sajmera-pensando — CHANGES_REQUESTED** [review]
Nice to see the upgrade/k0s invariant getting pinned down — the Raft-state-recovery and unit-adoption framing in the description is exactly right. Feedback below on how far the test actually reaches and what I'd want before we call zero-downtime-with-k0s validated. **WireGuard is out of scope here** (separate PR), so everything is the no-WG path only.

**What the test proves today**

Every assertion reads Spur's *own* control-plane view — `spur k8s status` (phase + `members:`) and `sinfo`. That legitimately covers: the controller's Raft-replicated `K0sClusterState` recovers across a `spurctld` restart, `spurd` re-adopts the running k0s unit instead of restarting it (`adopt_running_unit` in `crates/spurd/src/cluster.rs`), and nodes re-register. Good foundation.

**1. Main gap: no Kubernetes data-plane assertion**

The test never runs `kubectl`, never deploys a workload, never hits the API server (the "canary pod" from the commit message didn't make the final file). A pod could be evicted or etcd could lose a member mid-restart and the test still passes. I'd add:
- **A long-lived Deployment across the whole cycle** — assert `Ready` replicas stay constant *and* `restartCount == 0` through the controller restart and every agent restart. `restartCount` catches a pod that bounced and recovered.
- **etcd ground truth** — `cluster.py` already has `etcd_member_count()` / `wait_etcd_members()`. Assert quorum is intact at every step, not just that Spur's phase says `ready`. The phase is derived from systemd unit liveness ("a proxy for etcd quorum, not a direct etcd health check", per the note in `crates/spurctld/src/cluster_k8s.rs`), so the CLI can report healthy while etcd has actually wobbled — that divergence is the thing worth catching.
- **k8s still *schedules* after upgrade** — deploy a *new* pod post-upgrade and see it reach `Running`. The current post-upgrade check is a Spur batch job after `k8s down`, which says nothing about the k8s scheduler.

**2. Validate the full k0s state surface, on both CLI and metrics**

k0s state doesn't persist to Postgres (the accounting DB is jobs/fairshare only) — the two surfaces are `spur k8s status` (Raft-backed `K0sClusterState`) and the `/metrics/k8s` Prometheus endpoint. The test checks a subset of the first and none of the second. After each restart:
- **CLI:** assert not just phase + `members:` but the `control-plane:` list, per-node roles, and join-token metadata are unchanged (`k8s_control_planes()` / `k8s_active_controllers()` already exist).
- **Metrics:** `spur_k8s_cluster_phase`, `spur_k8s_cluster_up`, and `spur_k8s_nodes_by_role` on `/metrics/k8s` re-converge after a `spurctld` restart. The metric accumulator is rebuilt from Raft state + heartbeats, so this pins down that the rebuild is correct — a distinct failure mode from the Raft replay itself.

**3. Stale-read race**

Assertions fire immediately after `restart_controller()` returns, and that helper only waits for `sinfo` to list nodes — not for a fresh reconcile tick. The `ready` it reads may be the Raft-replayed value from *before* the new leader re-probes liveness. Re-poll via `wait_k8s_phase("ready")` (or the etcd check above) so you're asserting a freshly-derived value.

**4. Mirror the real upgrade orchestration**

`rolling_upgrade.yml` **drains** each agent (`DRAINED`/`DOWN`) before restarting `spurd`, then `RESUME`s — the test just kills+restarts the process. Two things to cover:
- **Drain a k0s member and confirm it stays in the cluster.** Worth being explicit because there are two unrelated "drains": `spur node drain` (scheduler drain, what the playbook uses) should leave k0s membership untouched — the unit is adopted and keeps running — whereas `spur k8s remove-nodes` (`drain_via_control_plane`, the kubectl cordon+drain+reset path) is the graceful *removal* path and must **not** be triggered by an upgrade drain. Test: `spur node drain` a k0s member → still a member, etcd unchanged, pods survive → `RESUME` → no k0s re-provision (role/mesh-IP identical, unit never restarted). The latent risk is whether the k0s reconcile treats a Spur-DRAINED node as unhealthy and strips its role (it shouldn't — the unit's still active — but that's exactly what a test should lock in).
- **Actual binary swap, not just a process restart.** As written this is a pure restart, so mixed-version behavior (controller upgraded, agents not yet) is untested. At minimum assert the version changed; ideally restart against a different build.

**5. Topology: cover k0s control-plane on a Spur worker node, and multi-CP / HA**

The fixture pins the CP to `node_names[0]`, which is also the `spurctld` host and the only controller. So it covers exactly one layout: a single k0s CP co-located with a single Spur controller. Worth covering:
- **k0s CP on a spurd-only worker node** — a different interleaving: a `spurctld` restart doesn't touch the k0s CP host, and unit re-adoption runs on a worker's `spurd`. This is arguably the more common production layout.
- **Multi-CP k0s (≥3, control planes spread across worker nodes)** — roll `spurd` through the CP nodes one at a time and assert `etcd_member_count() == 3` at every step. This is what proves adoption keeps etcd in quorum throughout, and it folds the worker-CP case in.
- **HA controller (≥3 `spurctld`)** — with a single controller, restarting it is a total reconcile blackout (no Raft leader). A 3-controller variant that rolls one at a time and asserts the k0s phase never leaves `ready` is what the docs mean by true zero-downtime.

**Comment by bhatturu** [comment]
@sajmera-pensando  addressed all your comments:

- Data-plane validation gap: deployed a canary Deployment (busybox sleep replicas) and assert restartCount==0 after each restart phase. Added etcd_member_count() stability check after controller restart and after rolling agent restart. New pod scheduling (post-upgrade-pod) proves the k8s scheduler is functional post-upgrade.
- State surface coverage: assert k8s_members(), k8s_control_planes(), and etcd_member_count() after controller restart and after rolling agent restart. Assert spur_k8s_cluster_phase{phase="ready"} and spur_k8s_cluster_up metrics via /metrics/k8s after each step.
- Stale-read race: added wait_k8s_phase("ready") re-poll after controller restart before any assertions fire, so assertions read freshly-reconciled Raft state instead of pre-restart values.
- Upgrade orchestration alignment: each agent restart is now preceded by spur node drain --reason rolling-upgrade, matching rolling_upgrade.yml play 4 drain semantics. Agent re-registers as idle after restart.
- Limited topology coverage: added test_rolling_upgrade_preserves_etcd_quorum to test_k8s_multicp.py. Brings up 3-CP k0s (--replicas 3), then rolls through controller restart + rolling drain+restart of all agents. Asserts 3-member etcd quorum, 3 active controllers, CP list stability, and metrics survive every step. Pure-controller topology (0 workers) skips canary/kubelet assertions since data-plane is already covered by test_upgrade_k0s.py.
- CI fix: k0s_down(reset=True) makes nodes non-responsive during teardown, causing Slurm to mark them down. Added explicit scontrol update State=RESUME for each node after agent re-registration so batch scheduling proceeds.

**Comment by sajmera-pensando** [comment]
Thanks for the updates — the data-plane canary + `restartCount`, etcd ground-truth via `etcd_member_count()`, the post-upgrade pod-scheduling check, the `wait_k8s_phase` re-poll for the stale-read race, and the `spur node drain` before each restart all land the core of my earlier feedback. A few gaps remain.

**Please address in this PR (each is a new test/assertion — no harness or CI changes needed):**

1. **`spur_k8s_nodes_by_role` metric** — you assert `spur_k8s_cluster_phase` and `spur_k8s_cluster_up` re-converge, but not `spur_k8s_nodes_by_role`. Snapshot it before the first restart and assert it's unchanged after the controller restart and after the rolling agent restarts. This is the metric that pins down the per-role accumulator rebuild.

2. **Per-node roles unchanged** — you check the `control-plane:` list, but not the per-node role rows from `spur k8s status`. Snapshot `{node: role}` and assert equal after each restart, so a role getting stripped/flipped is caught.

3. **Make the drain-vs-remove distinction explicit** — the drain path runs, but nothing asserts that a scheduler `spur node drain` does *not* cascade into a k0s removal (`drain_via_control_plane`). While a node is drained (before you restart its `spurd`), assert membership + `control-plane:` + `etcd_member_count()` + roles are all unchanged. That's the invariant worth locking in — the latent risk is the k0s reconcile treating a Spur-DRAINED node as unhealthy and stripping its role.

4. **k0s CP on a spurd-only worker** — the fixture pins the CP to `node_names[0]`, which is also the sole `spurctld` host, so only the co-located layout is covered. Add a test that brings k0s up with `--control-plane-node node_names[1]`, then restarts `spurctld` (node 0, which doesn't host the k0s CP) and restarts the CP-host `spurd` (node 1, where adoption actually runs), asserting the same invariants. This is arguably the more common production layout.

**Follow-up / not this PR (each needs surface that doesn't exist yet — a separate change, not a test):**

- **HA controller (≥3 `spurctld`):** the e2e harness only starts a single controller on node 0 (`_start_controller`); `test_controller_failover.py` already notes full multi-controller Raft is out of scope. A rolling-controller test needs multi-controller support in `cluster.py` first.
- **Join-token metadata unchanged:** `spur k8s status` deliberately never emits join tokens (bearer secrets), so there's no read-only surface to assert them against. Needs a CLI/RPC addition before it's testable.
- **Actual binary/version swap:** as discussed, fine to defer — this stays a pure process restart, so mixed-version behavior is out of scope here.

Copilot's L206 (k0s system images pre-`k8s up`) — agreed it's an infra precondition on cold VMs, not a test bug.

**Comment by bhatturu** [comment]
@sajmera-pensando addressed your latest comments in e556716:

- spur_k8s_nodes_by_role metric: snapshotted before the first restart, asserted unchanged after controller restart and after rolling agent restart.
- Per-node roles: snapshotted {node: role} from `spur k8s status`, asserted unchanged after controller restart, after each individual drain, and after rolling agent restart.
- Drain-vs-remove: `_drain_and_restart_agent` now asserts that while a node is drained, k8s membership, control-plane list, etcd quorum, and per-node roles are all unchanged — verifying drain does not cascade into a k0s removal.
- Worker-node CP topology: added `test_worker_node_cp_survives_restart` — brings k0s up with `--control-plane-node node_names[1]` (spurd-only worker, not co-located with spurctld). Asserts CP list, roles, membership, etcd, and metrics survive both controller restart and restart of the CP-node's agent.

**Review by sajmera-pensando — APPROVED** [review]
LGTM
